src/Shared/Debugger/HttpContextDebugFormatter.cs

```diff
@@ -53,6 +61,13 @@ public static string ContextToString(HttpContext context, string? reasonPhrase)
 
     private static string GetRequestUrl(HttpRequest request, bool includeQueryString)
     {
+        // The URL might be missing because the context was manually created in a test, e.g. new DefaultHttpContext()
+        // If there is no request URL then provide a more understandable empty value.
+        if (string.IsNullOrEmpty(request.Scheme) || !request.Host.HasValue)
+        {
+            return "(unset)";
+        }
+
         return $"{request.Scheme}://{request.Host.Value}{request.PathBase.Value}{request.Path.Value}{(includeQueryString ? request.QueryString.Value : string.Empty)}";
```

Comment:
How about always rendering the full url, but fill in missing scheme with `(unspecified)`. A missing host is ok.

`(unspecified):///path?query` or `(unspecified)://(unspecified)/path?query`

Reply:
Done.

I made it so that if everything is missing it just outputs `(unspecified)`. Having `(unspecified)://(unspecified)` everywhere felt too verbose.